src/cargo/core/registry.rs

```diff
@@ -718,3 +743,114 @@ fn lock(
+            return Ok((summary, Some(*lock_id)));
+        } else {
+            return Ok((summary, None));
+        }
+    }
+    assert!(summaries.is_empty());
+    // No summaries found, try to help the user figure out what is wrong.
+    if let Some((locked_patch, locked_id)) = locked {
+        // Since the locked patch did not match anything, try the unlocked one.
+        let mut orig_matches = match source.query_vec(orig_patch) {
+            Ok(summaries) => summaries,
+            Err(e) => {
+                log::warn!(
+                    "could not determine unlocked summaries for dep {:?}: {:?}",
+                    orig_patch,
+                    e
+                );
+                Vec::new()
+            }
+        };
+        if orig_matches.is_empty() {
+            // This should be relatively unusual. For example, a patch of
+            // {version="0.1.2", ...} and the patch location no longer contains a
+            // version that matches "0.1.2". It is unusual to explicitly write a
+            // version in the patch.
+            anyhow::bail!(
+                "The patch is locked to {} in Cargo.lock, but the version in the \
+                patch location does not match any packages in the patch location.\n\
+                Make sure the patch points to the correct version.",
+                locked_patch.version_req(),
+            );
+        }
+        let summary = best_summary(&mut orig_matches);
+        debug!(
+            "locked patch no longer matches, but unlocked version should work. \
+            locked={:?} unlocked={:?} summary={:?}",
+            locked, orig_patch, summary
+        );
+        // The unlocked version found a match. This returns a value to
+        // indicate that this entry should be unlocked.
+        return Ok((summary, Some(*locked_id)));
+    } else {
+        // Try checking if there are *any* packages that match this by name.
+        let name_only_dep =
+            Dependency::new_override(orig_patch.package_name(), orig_patch.source_id());
+        let found = match source.query_vec(&name_only_dep) {
+            Ok(name_summaries) => {
+                let mut vers = name_summaries
+                    .iter()
+                    .map(|summary| summary.version())
+                    .collect::<Vec<_>>();
+                match vers.len() {
+                    0 => format!(""),
+                    1 => format!("version `{}`", vers[0]),
+                    _ => {
+                        vers.sort();
+                        let strs: Vec<_> = vers.into_iter().map(|v| v.to_string()).collect();
+                        format!("versions `{}`", strs.join(", "))
+                    }
+                }
+            }
+            Err(e) => {
+                log::warn!(
+                    "failed to do name-only summary query for {:?}: {:?}",
+                    name_only_dep,
+                    e
+                );
+                "".to_string()
```

Comment:
This might be a bit simpler to explicitly ignore it with:

```rust
let found = source.query_vec(...).unwrap_or_else(|| {
    warn!(...); 
    Vec::new()
});
```